app/models/server/models/Users.js

```diff
@@ -204,6 +210,64 @@ export class Users extends Base {
 		return null;
 	}
 
+	// get next agent ignoring the ones reached the max amount of active chats
+	getAgentLoadByUsernames(usernames, customFilter) {
+		const usernameFilter = Array.isArray(usernames) ? { username: { $in: usernames } } : {};
+		const col = this.model.rawCollection();
+		return col.aggregate([
+			{
+				$match: {
+					...usernameFilter,
+					status: { $exists: true, $ne: 'offline' },
+					statusLivechat: 'available',
+					roles: 'livechat-agent',
+				},
+			},
+			{
+				$lookup: {
+					from: 'rocketchat_subscription',
+					localField: '_id',
+					foreignField: 'u._id',
+					as: 'subs',
+				},
+			},
+			{
+				$project: {
+					agentId: '$_id',
+					'livechat.maxNumberSimultaneousChat': 1,
+					username: 1,
+					lastAssignTime: 1,
+					lastRoutingTime: 1,
+					'queueInfo.chats': {
+						$size: {
+							$filter: {
+								input: '$subs',
+								as: 'sub',
+								cond: {
+									$and: [
+										{ $eq: ['$$sub.t', 'l'] },
+										{ $eq: ['$$sub.open', true] },
+										{ $ne: ['$$sub.onHold', true] },
+									],
+								},
+							},
+						},
+					},
+				},
+			},
+			...customFilter ? [customFilter] : [],
+			{
+				$sort: {
+					'queueInfo.chats': 1,
+					lastAssignTime: 1,
+					lastRoutingTime: 1,
+					username: 1,
+				},
+			},
+		]).toArray();
```

Comment:
Why not just return a list of usernames?
You are processing the list of users returned here anyway, which introduces another step in the stack.